lib/Driver/UnixToolChains.cpp

```diff
@@ -320,6 +322,50 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   return II;
 }
 
+
+ToolChain::InvocationInfo
+toolchains::GenericUnix::constructInvocation(const ArchiveJobAction &job,
+                               const JobContext &context) const {
+   assert(context.Output.getPrimaryOutputType() == file_types::TY_Image &&
+         "Invalid linker output type.");
+
+  ArgStringList Arguments;
+
+  // Configure the toolchain.
+  const char *AR = nullptr;
+  if (const Arg *A = context.Args.getLastArg(options::OPT_tools_directory)) {
+    StringRef toolchainPath(A->getValue());
+
+    // If there is an llvm-ar in the toolchain folder, use that instead.
+    if (auto toolchainAR =
+            llvm::sys::findProgramByName("llvm-ar", {toolchainPath})) {
+      AR = context.Args.MakeArgString(toolchainAR.get());
+    }
+  }
+
+  if (AR == nullptr) {
+    if (auto pathAR = llvm::sys::findProgramByName("ar", None))
+      AR = context.Args.MakeArgString(pathAR.get());
+  }
+
+  assert(AR &&
+         "neither ar nor llvm-ar was not found in the toolchain directory or system path.");
```

Comment:
This assert is poor - Release builds will drop it and you get no indication of why the command failed.  Please use a proper diagnostic here.

Reply:
We haven't been emitted diagnostics from *ToolChains.cpp. Maybe that should change, but most of the other jobs just fall back to invoking the bare name `"ar"` and getting the "not found" error from the system.